### PR TITLE
More memory leaks

### PR DESCRIPTION
**Pull request by jglick**
Cf. #41.

@reviewbybees

### CODE REVIEW

**Comment by abayer** [comment]
:bee:

**Comment by jglick** [comment]
I also want to check behavior with shared libraries (internal ~ `GroovyShellDecoratorImpl` as well as external ~ `ClasspathAdder`); and #59 ought to augment this test to check that Groovy step definitions are cleared as well.

**Comment by svanoort** [comment]
🐝

**Comment by jglick** [comment]
Hmm, I wonder if I could just use `GroovyClassLoader.getLoadedClasses()`…then I would merely need to enumerate the loaders. The main `GroovyClassLoader` inheritance chain is easy enough to follow; the tricky bit is that each compilation creates a new `InnerLoader`.

**Comment by jglick** [comment]
Hard leak has been fixed, but I wonder if we can also clear the soft references:

```
private static final org.codehaus.groovy.reflection.ClassInfo$ClassInfoSet org.codehaus.groovy.reflection.ClassInfo.globalClassSet->
org.codehaus.groovy.reflection.ClassInfo$ClassInfoSet@48d46169-segments->
[Lorg.codehaus.groovy.util.AbstractConcurrentMapBase$Segment;@5f8ffbd0-[10]->
org.codehaus.groovy.reflection.ClassInfo$ClassInfoSet$Segment@30e6148f-table->
[Ljava.lang.Object;@565a8dc4-[8]->
org.codehaus.groovy.reflection.ClassInfo@29f484c8-ref->
org.codehaus.groovy.util.ReferenceType$SoftRef@27734616-referent->
java.lang.Class@573d4844-<changed>->
groovy.lang.GroovyClassLoader$InnerLoader@6595381a
```

**Comment by rtyler** [comment]
@abayer Do you think these fixes could be part of the PermGen leak behavior [we](http://lists.jenkins-ci.org/pipermail/jenkins-infra/2016-September/000858.html) (and I know others) have seen on JDK7 with Multibranch?

**Comment by jglick** [comment]
This is heap, not permanent generation. And run Java 8!

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 1]
```diff
@@ -1019,6 +1022,23 @@ private static void cleanUpGlobalClassSet(@Nonnull Class<?> clazz) throws Except
         }
     }
 
+    private static void cleanUpObjectStreamClassCaches(@Nonnull Class<?> clazz) throws Exception {
+        Class<?> cachesC = Class.forName("java.io.ObjectStreamClass$Caches");
+        for (String cacheFName : new String[] {"localDescs", "reflectors"}) {
+            Field cacheF = cachesC.getDeclaredField(cacheFName);
+            cacheF.setAccessible(true);
+            ConcurrentMap<Reference<Class<?>>, ?> cache = (ConcurrentMap) cacheF.get(null);
+            Iterator<? extends Entry<Reference<Class<?>>, ?>> iterator = cache.entrySet().iterator();
```
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: How is mutation by concurrent threads handled wrt the iterator in this implementation?

**Comment by jglick** [comment]
Still seeing the test straining to collect loaders, with

```
private static org.codehaus.groovy.reflection.ClassInfo some.arbitrary.class.here.implemented.in.Groovy.$staticClassInfo->
org.codehaus.groovy.reflection.ClassInfo@3b4340f8-segment->
org.codehaus.groovy.reflection.ClassInfo$ClassInfoSet$Segment@391eea7f-table->
[Ljava.lang.Object;@fd26c3b-[28]->
[Ljava.lang.Object;@5855cc7b-[0]->
org.codehaus.groovy.reflection.ClassInfo@4af28a48-weakMetaClass->
org.codehaus.groovy.util.ManagedReference@219c59f9-ref->
org.codehaus.groovy.util.ReferenceType$SoftRef@1e723fcd-referent->
groovy.lang.MetaClassImpl@3c355255-metaMethodIndex->
org.codehaus.groovy.runtime.metaclass.MetaMethodIndex@19f9130d-table->
[Lorg.codehaus.groovy.runtime.metaclass.MetaMethodIndex$Entry;@1771a9cc-[84]->
org.codehaus.groovy.runtime.metaclass.MetaMethodIndex$Entry@16d07783-cachedStaticMethod->
org.codehaus.groovy.runtime.metaclass.MetaMethodIndex$CacheEntry@7b35f458-params->
[Ljava.lang.Class;@606ad775-[0]->
java.lang.Class@9168226-<changed>->
groovy.lang.GroovyClassLoader$InnerLoader@4142b833
```

and on a live instance, `ThreadGroupContext.beanInfoCache` seems to be holding on to a bunch of Pipeline Script classes even though we have called `Introspector.flushFromCaches`.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java (hunk 1 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Should be fine, it is a `ConcurrentHashMap`.

**Comment by jglick** [comment]
[JENKINS-32338](https://issues.jenkins-ci.org/browse/JENKINS-32338) seems to report something similar.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java [hunk 2]
```diff
@@ -362,6 +362,7 @@ private boolean run() {
         if (ending) {
             execution.cleanUpHeap();
             scripts.clear();
+            closures.clear();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java]: Actually the bug was that it was nonempty at this point. Will correct in #279.

**Pull request merged**